### PR TITLE
Add type hints to all functions in metrics.py #446

### LINKED ISSUE #446: Add type hints to functions in metrics.py
## Summary
All functions in `neural_lam/metrics.py` currently lack type annotations. 
This makes IDE support incomplete and mypy (already configured in 
pre-commit) cannot fully validate the module.

## Functions affected
- `get_metric`
- `mask_and_reduce_metric`
- `wmse`, `mse`, `wmae`, `mae`, `nll`, `crps_gauss`

## Proposed Fix
Add `torch.Tensor` and `Optional[torch.Tensor]` type hints to all function 
signatures and return types.

## Why This Matters
- mypy is already set up in `.pre-commit-config.yaml`
- Type hints improve contributor experience and catch bugs early
- Consistent with type hints already present in `loss_weighting.py`

### PR DESCRIPTION
**Pull request by sidhantpande**
## Describe your changes
Add `torch.Tensor`, `Optional[torch.Tensor]` and `bool` type hints to all
function signatures and return types in `neural_lam/metrics.py`.

Functions updated:
- `get_metric`
- `mask_and_reduce_metric`
- `wmse`, `mse`, `wmae`, `mae`, `nll`, `crps_gauss`

No logic changes were made — only type annotations were added.

## Issue Link
closes #446

## Type of change
- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review
- [x] My branch is up-to-date with the target branch
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the README to cover introduced code changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form
- [ ] I have requested a reviewer and an assignee

## Checklist for reviewers
Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [ ] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review
- [ ] I have added a line to the CHANGELOG describing this change:
  - *maintenance*: add type hints to all functions in `metrics.py`

## Checklist for assignee
- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: neural_lam/metrics.py [hunk 1]
```diff
@@ -1,8 +1,11 @@
+# Standard library
+from typing import Optional
+
 # Third-party
 import torch
 
 
-def get_metric(metric_name):
+def get_metric(metric_name: str):
```
Comment (sadamov) [inline neural_lam/metrics.py]: `get_metric` is missing a return type. The PR description says "all functions" but this one's incomplete. Since it returns one of the metric callables, something like this works:

```suggestion
def get_metric(metric_name: str) -> "Callable[..., torch.Tensor]":
```
You'd also need `from collections.abc import Callable` (or `from typing import Callable`) added to the imports.

**Review by sadamov — COMMENTED** [review]
Overall looks good, just a few small things.

File: neural_lam/metrics.py [hunk 2]
```diff
@@ -1,8 +1,11 @@
+# Standard library
+from typing import Optional
```
Comment (sadamov) [inline neural_lam/metrics.py:3]: Project requires Python >= 3.10, so `Optional` from `typing` is a legacy alias. `torch.Tensor | None` is preferred. If you update the usages, this import can be dropped entirely.

```suggestion
from collections.abc import Callable
```
(if you also fix the `get_metric` return type above)

**Comment by sadamov** [comment]
CHANGELOG entry is missing

**Review by sadamov — APPROVED** [review]
Thanks @sidhantpande for the hints, sorry for the long wait! I've rebased onto current main and applied the review feedback: get_metric now returns Callable[..., torch.Tensor], and I kept Optional from typing rather than | None to stay consistent with the other typehint PRs. Merging after CI is green.

**Review by sadamov — APPROVED** [review]
Thanks @sidhantpande for the hints, sorry for the long wait! I've rebased onto current main and applied the review feedback: get_metric now returns Callable[..., torch.Tensor], and I kept Optional from typing rather than | None to stay consistent with the other typehint PRs. Merging after CI is green.
